controllers/replication_controller.go

```diff
@@ -132,78 +104,189 @@ func (r *ReconcileReplication) Reconcile(request reconcile.Request) (reconcile.R
+	}
+
+	if execution.InProgress > 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// reconcileFinishedReplicationExecution fetches the latest finished replication execution and returns an error when it has failed.
+func (r *ReplicationReconciler) reconcileFinishedReplicationExecution(ctx context.Context, replication *registriesv1alpha1.Replication, harborClient *h.RESTClient) error {
+	replExec := &legacymodel.ReplicationExecution{
+		PolicyID: replication.Status.ID,
+		Trigger:  registriesv1alpha1.ReplicationTriggerTypeManual,
+	}
+
+	executions, err := harborClient.GetReplicationExecutions(ctx, replExec)
+	if err != nil {
+		return fmt.Errorf("fetching successful replication executions failed for replication policy %s: %s", replication.Status.Name, err)
+	}
+
+	newestReplExecID := getNewestReplicationExecutionID(executions)
+
+	execution, err := harborClient.GetReplicationExecutionByID(ctx, newestReplExecID)
+	if err != nil {
+		return fmt.Errorf("could not get the latest finished replication execution, id: %d, error: %s", newestReplExecID, err)
+	}
+
+	if execution.Failed > 0 || execution.Status == "Failed" {
+		return fmt.Errorf("latest execution was not successful, replication policy: %s", replication.Spec.Name)
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReplicationReconciler implements reconcile.Reconciler.
+var _ reconcile.Reconciler = &ReplicationReconciler{}
+
+func (r *ReplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Create a new controller
+	c, err := controller.New("replication-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Replication
+	err = c.Watch(&source.Kind{Type: &registriesv1alpha1.Replication{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getNewestReplicationExecutionID takes a slice of replication executions and returns the one with the highest ID.
+func getNewestReplicationExecutionID(executions []*legacymodel.ReplicationExecution) int64 {
+	max := executions[0].ID
+	for i := range executions {
+		if max < executions[i].ID {
+			max = executions[i].ID
+		}
+	}
+	return max
+}
+
 // updateReplicationCR compares the new CR status and finalizers with the pre-existing ones and updates them accordingly
-func (r *ReconcileReplication) updateReplicationCR(ctx context.Context, parentInstance *registriesv1alpha1.Instance,
-	originalReplication, replication *registriesv1alpha1.Replication) (reconcile.Result, error) {
+func (r *ReplicationReconciler) updateReplicationCR(ctx context.Context, parentInstance *registriesv1alpha1.Instance,
+	originalReplication, replication *registriesv1alpha1.Replication) (ctrl.Result, error) {
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Name:      originalReplication.Name,
+		Namespace: originalReplication.Namespace,
+	},
+		originalReplication); err != nil {
```

Comment:
Why pass `originalReplication` as a parameter, if it is immediately replaced by a `Client.Get`?